File: isegoria/src/main/java/com/eulersbridge/isegoria/poll/PollVoteFragment.java
package com.eulersbridge.isegoria.poll;

import android.arch.lifecycle.ViewModelProviders;
import android.os.Bundle;
import android.support.annotation.NonNull;
import android.support.v4.app.Fragment;
import android.support.v7.widget.RecyclerView;
import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;
import android.widget.ImageView;
import android.widget.TextView;

import com.bumptech.glide.load.resource.drawable.DrawableTransitionOptions;
import com.eulersbridge.isegoria.GlideApp;
import com.eulersbridge.isegoria.R;
import com.eulersbridge.isegoria.network.api.models.Poll;
import com.eulersbridge.isegoria.util.Constants;

import org.parceler.Parcels;

public class PollVoteFragment extends Fragment implements PollOptionAdapter.PollOptionVoteListener {

    private PollViewModel viewModel;

    private ImageView creatorImageView;
    private TextView questionTextView;
    private TextView creatorNameTextView;
    private TextView answersCountTextView;

    private PollOptionAdapter adapter;
	
	@Override
	public View onCreateView(@NonNull LayoutInflater inflater, ViewGroup container, Bundle savedInstanceState) {

        View rootView = inflater.inflate(R.layout.poll_fragment, container, false);

        viewModel = ViewModelProviders.of(this).get(PollViewModel.class);
        setupViewModelObservers();

        creatorImageView = rootView.findViewById(R.id.poll_vote_creator_image_view);
        questionTextView = rootView.findViewById(R.id.poll_vote_question_text_view);
        creatorNameTextView = rootView.findViewById(R.id.poll_vote_creator_name_text_view);
        answersCountTextView = rootView.findViewById(R.id.poll_vote_answers_count_text_view);

        adapter = new PollOptionAdapter(this);

        RecyclerView listView = rootView.findViewById(R.id.poll_fragment_list_view);
        listView.setAdapter(adapter);

        Poll poll = Parcels.unwrap(getArguments().getParcelable(Constants.ACTIVITY_EXTRA_POLL));
        viewModel.setPoll(poll);

		return rootView;
	}

	private void setupViewModelObservers() {
	    viewModel.poll.observe(this, poll -> {
	        if (poll != null) {
                questionTextView.setText(poll.question);

                populatePollOptions();
            }
        });

	    viewModel.pollCreator.observe(this, creator -> {
	        if (creator != null) {
                GlideApp.with(this)
                        .load(creator.profilePhotoURL)
                        .placeholder(R.color.lightGrey)
                        .transition(DrawableTransitionOptions.withCrossFade())
                        .into(creatorImageView);

                creatorNameTextView.setText(getString(R.string.poll_asked_by, creator.getFullName()));
                creatorNameTextView.setVisibility(View.VISIBLE);
            }
        });

	    viewModel.pollResults.observe(this, results -> {
	        if (results == null) return;

            int answersCount = results.size();

            answersCountTextView.setText(String.valueOf(answersCount));
            answersCountTextView.setContentDescription(getString(R.string.poll_vote_answers_content_description, answersCount));

            populatePollOptions();
        });
    }

    private void populatePollOptions() {
	    Poll poll = viewModel.poll.getValue();

        if (poll != null && poll.options != null) {
            adapter.setLoading(false);
            adapter.replaceItems(poll.options);
            adapter.setPollVotingEnabled(!poll.closed);
        }
    }

    @Override
    public void onPollOptionClick(int position) {
	    viewModel.voteForPollOption(position);
    }
}

File: isegoria/src/main/kotlin/com/eulersbridge/isegoria/poll/PollVoteFragment.kt
package com.eulersbridge.isegoria.poll

import android.arch.lifecycle.ViewModelProvider
import android.arch.lifecycle.ViewModelProviders
import android.content.Context
import android.os.Bundle
import android.support.v4.app.Fragment
import android.view.LayoutInflater
import android.view.View
import android.view.ViewGroup
import com.bumptech.glide.load.resource.drawable.DrawableTransitionOptions
import com.eulersbridge.isegoria.GlideApp
import com.eulersbridge.isegoria.MainActivity
import com.eulersbridge.isegoria.R
import com.eulersbridge.isegoria.util.extension.ifTrue
import com.eulersbridge.isegoria.util.extension.observe
import com.eulersbridge.isegoria.util.extension.observeBoolean
import com.eulersbridge.isegoria.util.extension.setCompatTooltipText
import dagger.android.support.AndroidSupportInjection
import kotlinx.android.synthetic.main.poll_fragment.*
import javax.inject.Inject

class PollVoteFragment : Fragment(), PollOptionAdapter.PollOptionVoteListener {

    @Inject
    internal lateinit var modelFactory: ViewModelProvider.Factory
    private lateinit var viewModel: PollVoteViewModel

    private var pollOptionsAdapter: PollOptionAdapter = PollOptionAdapter(this)

    override fun onAttach(context: Context?) {
        AndroidSupportInjection.inject(this)
        super.onAttach(context)
        viewModel = ViewModelProviders.of(this, modelFactory)[PollVoteViewModel::class.java]
    }

    override fun onCreateView(
        inflater: LayoutInflater,
        container: ViewGroup?,
        savedInstanceState: Bundle?
    ): View? {

        val rootView = inflater.inflate(R.layout.poll_fragment, container, false)

        viewModel.setPoll(arguments?.getParcelable(FRAGMENT_EXTRA_POLL))

        return rootView
    }

    override fun onViewCreated(view: View, savedInstanceState: Bundle?) {
        listView.adapter = pollOptionsAdapter

        createViewModelObservers()
    }

    private fun createViewModelObservers() {
        observe(viewModel.pollQuestion) { questionTextView.text = it!! }
        observe(viewModel.pollOptions) { pollOptionsAdapter.replaceItems(it!!) }
        observeBoolean(viewModel.votingEnabled) { pollOptionsAdapter.pollVotingEnabled = it }

        observe(viewModel.pollCreator) {
            if (it != null) {
                GlideApp.with(this)
                        .load(it.profilePhotoURL)
                        .placeholder(R.color.lightGrey)
                        .transition(DrawableTransitionOptions.withCrossFade())
                        .into(creatorImageView)

                creatorNameTextView.apply {
                    text = getString(R.string.poll_asked_by, it.fullName)
                    visibility = View.VISIBLE
                }
            }
        }

        observe(viewModel.pollResults) { results ->
            results?.let {
                val answerCount = results.size

                answersCountTextView.text = answerCount.toString()

                val answersQuantity =
                    resources.getQuantityString(R.plurals.poll_vote_answers_content_description, answerCount)
                answersCountTextView.contentDescription = answersQuantity
                answersCountTextView.setCompatTooltipText(answersQuantity)
            }
        }

        ifTrue(viewModel.votingError) {
            (activity as? MainActivity)?.showSnackbar(R.string.poll_vote_failed)
        }
    }

    override fun onPollOptionClick(position: Int) {
        viewModel.voteForPollOption(position)
    }
}
